Signal/src/views/DirectionalPanGestureRecognizer.swift

```diff
@@ -17,19 +16,50 @@ class PanDirectionGestureRecognizer: UIPanGestureRecognizer {
 
     init(direction: PanDirection, target: AnyObject, action: Selector) {
         self.direction = direction
+
         super.init(target: target, action: action)
     }
 
     override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent) {
+        // Only start gesture if it's initially in the specified direction.
+        if state == .possible {
+            guard let touch = touches.first else {
+                return
+            }
+
+            let previousLocation = touch.previousLocation(in: view)
+            let location = touch.location(in: view)
+            let deltaY = previousLocation.y - location.y
+            let deltaX = previousLocation.x - location.x
+
+            switch direction {
+            case .down where deltaY < 0:
+                return
+            case .up where deltaY > 0:
+                return
+            case .forward where deltaX < 0:
+                return
+            case .backward where deltaX > 0:
+                return
+            default:
+                break
+            }
+        }
+
+        // Gesture was already started, or in the correct direction.
         super.touchesMoved(touches, with: event)
 
         if state == .began {
             let vel = velocity(in: view)
             switch direction {
-            case .horizontal where fabs(vel.y) > fabs(vel.x):
-                state = .cancelled
-            case .vertical where fabs(vel.x) > fabs(vel.y):
-                state = .cancelled
+            case .forward, .backward:
+                if fabs(vel.y) > fabs(vel.x) {
+                    state = .cancelled
```

Comment:
`UIGestureRecognizer`s can't always be cancelled; IMO safer to have an `abortGesture()` method that not only _tries_ to enter cancelled state but also toggles the enabled/disabled state, thereby forcing the gesture to abort.

Reply:
When can't they be cancelled?

Reply:
It's a particularity of `UIGestureRecognizer` that cancelling doesn't always work.  I can't remember the exact repro, but it's not terribly obscure.